apps/src/code-studio/pd/workshop_dashboard/workshops/overview/sections/WorkshopStatusSection.tsx

```diff
@@ -0,0 +1,354 @@
+    },
+  },
+  {
+    stateKey: 'reopen',
+    label: 'Reopen',
+    description:
+      'Are you sure you want to reopen this workshop and change it back to "In Progress"? Note reopening then ending again will send exit survey emails for new attendees, but will not re-send surveys that were already sent.',
+    primaryButtonProps: {},
+  },
+  {
+    stateKey: 'unstart',
+    label: 'Unstart',
+    description:
+      'Are you sure you want to unstart this workshop and change it back to "Not Started?"',
+    primaryButtonProps: {},
+  },
+];
+
+export type WorkshopActions = 'start' | 'end' | 'unstart' | 'reopen';
+
+export const WorkshopStatusSection: React.FC<WorkshopStatusSectionProps> = ({
+  workshop,
+  isWorkshopAdmin,
+  onWorkshopUpdate,
+}) => {
+  const [activeDialog, setActiveDialog] = useState<WorkshopActions | null>(
+    null
+  );
+  const [isUpdating, setIsUpdating] = useState(false);
+  const [error, setError] = useState<string | null>(null);
+
+  const handleClick = (stateKey: WorkshopActions) => {
+    setError(null);
+    setActiveDialog(stateKey);
+  };
+
+  const generateHandler = useCallback(
+    (action: WorkshopActions) => async () => {
+      setIsUpdating(true);
+      setError(null); // Clear any previous errors
+      try {
+        const response = await fetch(
+          `/api/v1/pd/workshops/${workshop.id}/${action}`,
+          {
+            method: 'POST',
+            headers: {
+              'Content-Type': 'application/json',
+              'X-CSRF-Token': await getAuthenticityToken(),
+            },
+          }
+        );
+
+        if (response.ok) {
+          onWorkshopUpdate();
+        } else {
+          let errorMessage = `Failed to ${action} workshop. Please try again.`;
+          const errorData = await response.json();
+          if (errorData.error) {
+            errorMessage = errorData.error;
+          } else if (
+            errorData.errors &&
+            Array.isArray(errorData.errors) &&
+            errorData.errors.length > 0
+          ) {
+            errorMessage = errorData.errors.join(', ');
+          }
+
+          if (response.status === 403) {
+            errorMessage = 'You do not have permission to perform this action.';
+          } else if (response.status === 404) {
+            errorMessage = 'Workshop not found.';
+          } else if (response.status >= 500) {
+            errorMessage = 'Server error occurred. Please try again later.';
+          }
+
+          setError(errorMessage);
+        }
+      } catch (unknownError) {
+        setError('An unknown error occurred. Please try again.');
+      } finally {
+        setIsUpdating(false);
+        setActiveDialog(null);
+      }
+    },
+    [onWorkshopUpdate, workshop.id]
+  );
+
+  const notStarted = workshop.state === 'Not Started';
+  const inProgress = workshop.state === 'In Progress';
+  const ended = workshop.state === 'Ended';
+  const cannotEndWorkshop = !workshop.readyToClose;
```

Comment:
Nit: Would it be possible to have the constants all together at the top of the component?